src/vm-vcpu-ref/src/x86_64/gdt.rs

```diff
@@ -1,114 +1,254 @@
+    /// Build a `kvm_segment` from a GDT segment descriptor and the
+    /// selector (`table_index`).
+    ///
+    /// # Arguments
+    ///
+    /// * `entry` - The gdt entry.
+    /// * `table_index` - Index of the entry in the gdt table.
+    ///
+    /// # Example - Create the kvm_segment corresponding to the code
+    ///
+    /// ```rust
+    /// use vm_vcpu_ref::x86_64::gdt::SegmentDescriptor;
+    ///
+    /// let code_descriptor = SegmentDescriptor::from(0xa09b, 0, 0xfffff);
+    /// // The code descriptor is found at position 1 in the GDT.
+    /// // More examples are available here: https://wiki.osdev.org/GDT_Tutorial
+    /// let table_index = 1;
+    /// let code_segment = code_descriptor.create_kvm_segment(table_index);
+    /// ```
+    pub fn create_kvm_segment(&self, table_index: u8) -> kvm_segment {
+        kvm_segment {
+            base: self.base(),
+            limit: self.limit(),
+            selector: u16::from(table_index * 8),
+            type_: self.segment_type(),
+            present: self.p(),
+            dpl: self.dpl(),
+            db: self.db(),
+            s: self.s(),
+            l: self.l(),
+            g: self.g(),
+            avl: self.avl(),
+            padding: 0,
+            unusable: match self.p() {
+                0 => 1,
+                _ => 0,
+            },
+        }
+    }
 }
 
-pub fn get_s(entry: u64) -> u8 {
-    ((entry & 0x0000_1000_0000_0000) >> 44) as u8
+#[derive(Clone, Debug)]
+/// The `Gdt` is a wrapper for creating and managing operations on the
+/// Global Descriptor Table (GDT). The GDT is a data structure used by
+/// Intel x86-family processors to define the characteristics of the
+/// various memory areas used during program execution (like data, code, TSS).
+///
+/// The `Gdt` provides a default implementation that can be used for setting
+/// up a vCPU for booting. The default implementation contains all 4 segment
+/// descriptors corresponding to null, code, data, and TSS. The default can be
+/// extended by using `push`. When the default is not matching the product
+/// requirements, the GDT can also be created from scratch.
+pub struct Gdt(Vec<SegmentDescriptor>);
+
+impl Gdt {
+    /// Create an empty `GDT`.
+    ///
+    /// # Example - Creating a GDT from scratch
+    ///
+    /// ```rust
+    /// use vm_vcpu_ref::x86_64::gdt::{Gdt, SegmentDescriptor};
+    ///
+    /// let mut gdt = Gdt::new();
+    /// // Create the GDT as in the GDT Tutorial: https://wiki.osdev.org/GDT_Tutorial
+    /// gdt.push(SegmentDescriptor::from(0, 0, 0));
+    /// gdt.push(SegmentDescriptor::from(0x9A, 0, 0xffffffff));
+    /// gdt.push(SegmentDescriptor::from(0x92, 0, 0xffffffff));
+    /// gdt.push(SegmentDescriptor::from(0x89, 0, 0xffffffff));
+    /// ```
+    pub fn new() -> Gdt {
+        Gdt(vec![])
+    }
+
+    /// Writes the GDT (Global Descriptor Table) into Guest Memory.
+    ///
+    /// # Example - Creating the GDT for Linux
```

Comment:
For this one I find it superfluous to have the arguments section because it would just repeat what the documentation is already saying. I excluded it on purpose, but I can add it if you feel strongly about it.